File: backend/app/schemas/download.py
"""Request and response models for persistent download jobs."""

from typing import Literal
from datetime import datetime

from pydantic import BaseModel, Field

FormatPreset = Literal["best", "1080p", "720p", "480p", "audio"]


class CreateDownloadRequest(BaseModel):
    url: str = Field(min_length=1, max_length=2048)
    output_directory: str = Field(default=".", min_length=1, max_length=2048)
    format_preset: FormatPreset = "best"
    video_ids: list[str] | None = None


class DownloadItemResponse(BaseModel):
    id: int
    video_id: str | None
    title: str
    playlist_index: int | None
    status: str
    progress: float
    downloaded_bytes: int
    total_bytes: int | None
    speed: float | None
    eta: float | None
    output_path: str | None
    error_message: str | None
    created_at: datetime
    updated_at: datetime


class DownloadJobResponse(BaseModel):
    id: str
    playlist_title: str | None
    status: str
    total_items: int
    completed_items: int
    failed_items: int
    output_directory: str
    format_preset: str
    created_at: datetime
    started_at: datetime | None
    finished_at: datetime | None
    error_message: str | None
    items: list[DownloadItemResponse]

File: frontend/src/types/download.ts
export type FormatPreset = 'best' | '1080p' | '720p' | '480p' | 'audio'
export type JobStatus = 'queued' | 'inspecting' | 'downloading' | 'postprocessing' | 'completed' | 'failed' | 'cancelled' | 'interrupted'

export interface CreateDownload {
  url: string
  video_ids: string[]
  format_preset: FormatPreset
  output_directory: string
}

export interface DownloadItem {
  id: number
  video_id: string | null
  title: string
  status: JobStatus | 'skipped'
  progress: number
  downloaded_bytes: number
  total_bytes: number | null
  speed: number | null
  eta: number | null
  error_message: string | null
}

export interface DownloadJob {
  id: string
  playlist_title: string | null
  status: JobStatus
  total_items: number
  completed_items: number
  failed_items: number
  output_directory: string
  format_preset: FormatPreset
  error_message: string | null
  items: DownloadItem[]
}

export const isTerminal = (status: JobStatus) => ['completed', 'failed', 'cancelled', 'interrupted'].includes(status)
